hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobUtils.java

```diff
@@ -907,6 +789,143 @@ public static boolean hasMobColumns(TableDescriptor htd) {
+        fams.add(hcd);
+      }
+    }
+    return fams;
+  }
+
+  /**
+   * Performs housekeeping file cleaning (called by MOB Cleaner chore)
+   * @param conf configuration
+   * @param table table name
+   * @throws IOException
+   */
+  public static void cleanupObsoleteMobFiles(Configuration conf, TableName table)
+      throws IOException {
+
+    try (final Connection conn = ConnectionFactory.createConnection(conf);
+        final Admin admin = conn.getAdmin();) {
+      TableDescriptor htd = admin.getDescriptor(table);
+      List<ColumnFamilyDescriptor> list = getMobColumnFamilies(htd);
+      if (list.size() == 0) {
+        LOG.info("Skipping non-MOB table [" + table + "]");
+        return;
+      }
+      Path rootDir = FSUtils.getRootDir(conf);
+      Path tableDir = FSUtils.getTableDir(rootDir, table);
+      // How safe is this call?
+      List<Path> regionDirs = FSUtils.getRegionDirs(FileSystem.get(conf), tableDir);
+
+      Set<String> allActiveMobFileName = new HashSet<String>();
+      FileSystem fs = FileSystem.get(conf);
+      for (Path regionPath: regionDirs) {
+        for (ColumnFamilyDescriptor hcd: list) {
+          String family = hcd.getNameAsString();
+          Path storePath = new Path(regionPath, family);
+          boolean succeed = false;
+          Set<String> regionMobs = new HashSet<String>();
+          while(!succeed) {
+            //TODO handle FNFE
+            RemoteIterator<LocatedFileStatus> rit = fs.listLocatedStatus(storePath);
+            List<Path> storeFiles = new ArrayList<Path>();
+            // Load list of store files first
+            while(rit.hasNext()) {
+              Path p = rit.next().getPath();
+              if (fs.isFile(p)) {
+                storeFiles.add(p);
+              }
+            }
+            try {
+              for(Path pp: storeFiles) {
+                HStoreFile sf = new HStoreFile(fs, pp, conf, CacheConfig.DISABLED,
+                  BloomType.NONE, true);
+                sf.initReader();
+                byte[] mobRefData = sf.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+                byte[] mobCellCountData = sf.getMetadataValue(HStoreFile.MOB_CELLS_COUNT);
+                byte[] bulkloadMarkerData = sf.getMetadataValue(HStoreFile.BULKLOAD_TASK_KEY);
+                if (mobRefData == null && (mobCellCountData != null ||
+                    bulkloadMarkerData == null)) {
+                  LOG.info("Found old store file with no MOB_FILE_REFS: " + pp
+                    +" - can not proceed until all old files will be MOB-compacted");
+                  return;
+                } else if (mobRefData == null) {
+                  LOG.info("Skipping file without MOB references (can be bulkloaded file):"+ pp);
+                  continue;
+                }
+                String[] mobs = new String(mobRefData).split(",");
+                regionMobs.addAll(Arrays.asList(mobs));
+              }
+            } catch (FileNotFoundException e) {
```

Comment:
Yes, still need to analyze if it is a potential source of a data loss.

Reply:
Files can deleted (archived) only after compaction. As since, HBase can not run more than one compaction on a region at the same time and this code (FNFE) is executed during compaction, I do not see how it is possible to catch this FNFE in a first place. Looks safe to me.

Reply:
Only if someone deliberately deletes this file. I think we need abort and log FATAL error in this case.

Reply:
That was a wrong comment. The code is executed in a MOB file cleaner.

Reply:
Verified, that code is correct. On FNFE we start the iteration loop from very beginning, therefore, cleaning chore will succeed only when no FNFE will be thrown during execution. Safe.